contracts/roulette/lib.rs

```diff
@@ -0,0 +1,388 @@
+#![cfg_attr(not(feature = "std"), no_std, no_main)]
+
+/// The Roulette
+/// - there is a window of length N blocks for users to place their bets
+/// - there are M bets allowed in each such block
+/// - after that no more bets can be placed until spin is called and the winnings are paid out
+#[ink::contract]
+mod roulette {
+
+    #[cfg(feature = "std")]
+    use ink::storage::traits::StorageLayout;
+    use ink::{
+        codegen::EmitEvent,
+        env::{
+            call::{build_call, ExecutionInput},
+            set_code_hash, DefaultEnvironment, Error as InkEnvError,
+        },
+        prelude::{format, string::String},
+        reflect::ContractEventBase,
+        storage::{traits::ManualKey, Lazy, Mapping},
+    };
+    use scale::{Decode, Encode};
+
+    #[derive(Debug, PartialEq, Eq, Encode, Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum RouletteError {
+        InkEnvError(String),
+        ArithmethicError,
+        BetAmountIsTooSmall,
+        PlayerAlreadyPlacedABet,
+        NoMoreBetsCanBeMade,
+        BettingPeriodNotOver,
+        NativeTransferFailed(String),
+        NotEnoughBalance,
+        CallerIsNotTheHouseOwner,
+    }
+
+    impl From<InkEnvError> for RouletteError {
+        fn from(e: InkEnvError) -> Self {
+            RouletteError::InkEnvError(format!("{e:?}"))
+        }
+    }
+
+    pub type Selector = [u8; 4];
+    pub type Result<T> = core::result::Result<T, RouletteError>;
+    pub type Event = <Roulette as ContractEventBase>::Type;
+
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct BetPlaced {
+        #[ink(topic)]
+        player: AccountId,
+        #[ink(topic)]
+        bet_type: BetType,
+        amount: Balance,
+    }
+
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct WheelSpin {
+        winning_number: u8,
+    }
+
+    #[derive(Debug, Encode, Decode, Clone, Copy, PartialEq, Eq)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo, StorageLayout))]
+    pub enum BetType {
+        Number(u8),
+        Red,   // even
+        Black, // red
```

Comment:
{Red, Black} are not the same as {Even, Odd}. I messed this up too and actually got stuck on it for quite a long time. This is our [official double-zero table layout](https://bfgblog-a.akamaihd.net/uploads/2013/11/2-1-Roulette-Table-Wheel-1024x463.png). I included this link in the solidity solution and we should include it here too.

Reply:
added even / odd bets